### PR TITLE
Develop/directory structure

### PR DESCRIPTION
**Pull request by giangiac**
Files from /qureg, /util, /interface have been moved to /include, /src.
The distinction is based on the files being .h/.hpp or .cpp respectively.

Some files from /examples have been moved to /tutorials.

CMakeLists.txt has been updated and the build:
CXX=mpiicpc cmake -DIqsMPI=ON -DIqsUtest=ON -DIqsPython=OFF ..
was successfully checked.

### CODE REVIEW

**Comment by giangiac** [comment]
Fabio, the conflicts appear due to the last commit of the CPhase gate. They can be easily solved.

File: docs/directory_structure.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+|   |-- suite_of_tests.cpp
+|   |-- \include
+|   |   |-- <header files with unit tests>
+|   |-- \data
+|       |-- <files used in testing>
+|
+|-- \examples
+    |-- <C++ main files of small applications>
+    |-- <iPython notebooks of small applications>
+```
```
Comment (fbaru-dev) [inline docs/directory_structure.md:41]: I would add also a Cmakefile for each directory, in the tutorials and examples. In this way users can see how to link the code as an external library. We can do it later

Comment (fbaru-dev) [reply docs/directory_structure.md:41]: maybe the utility functions can be added in directory called utils, like ICHEC did or other projects

File: include/interface_api_qasm.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * Handle the execution of a QASM instruction in qHiPSTER.
+ * @param op The QASM operation we wish to execute.
+ * @param args The operands, if any, for the operation.
+ * @return 0 Success - Op Handled\n
+ *        >0 Failure - Op not handled or error.
+ */
+unsigned long ExecuteHandler(std::string op, std::string args);
+
+/** @}*/
```
Comment (fbaru-dev) [inline include/interface_api_qasm.h:27]: Should we rename it as .hpp file to be consistent?

File: include/interface_api_qubitid.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * query_qubit_id()
+ * @param qubit_name The string name passed from the QASM instruction.
+ * @return An integer representing the the qubit identifier that matches the
+ *         qubit name string. If the qubit name is not found then an id is generated, 
+ *         added to an internal tracking table, and the id is returned.
+ */
+int query_qubit_id(std::string qubit_name);
+
+/** @}*/
```
Comment (fbaru-dev) [inline include/interface_api_qubitid.h:25]: Should we rename it as .hpp file to be consistent?

File: include/interface_api_version.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+#define INTERFACE_VERSION_STRING "RC-1.1.0"
+
+/**
+ * Print out the QASM interface version string. The format will be 
+ * major.minor.revision.
+ * @param args Unused.
+ * @return 0.
+ */
+unsigned long quiversion(std::string args);
+
```
Comment (fbaru-dev) [inline include/interface_api_version.h:25]: Should we rename it as .hpp file to be consistent?

**Review by fbaru-dev — COMMENTED** [review]
Sure things, the conflict can be easily resolved

**Comment by fbaru-dev** [comment]
Merge successfully done!

**Pull request merged**